ui/puzzle/src/ctrl.ts

```diff
@@ -334,14 +334,18 @@ export default function (opts: PuzzleOpts, redraw: Redraw): Controller {
     if (streak && vm.lastFeedback != 'win') return;
 
     ceval.stop();
-    vm.next.promise.then(initiate).then(redraw);
+    vm.next.promise.then(initiate).then(redraw).catch(redirectToDashboard);
```

Comment:
not a fan of using a failed promise like this

Reply:
Yeah, I wasn't sure about that part. The dream to me would be:

`vm.next.promise.then(redirectIfUndefined).then(initiate).then(redraw);`

where `vm.next.promise: Deferred<PuzzleData | undefined>` and

```ts
function redirectIfUndefined(fromData: PuzzleData | undefined): PuzzleData {
  return fromData || lichess.redirect(...);
}
```

where `lichess.redirect` returns `never`. So then the types show what's going on. I don't think redirect can be thought of as returning never though.

I'll test out some alternate ideas.